File: lib/content_converter.py
"""Convert Confluence storage format XHTML to Flare-compatible HTML.

Handles Confluence-specific elements (ac:image, ac:structured-macro,
ac:layout, ri:attachment, etc.) and converts them to clean HTML that
works in MadCap Flare topics.
"""

import re
import logging

from bs4 import BeautifulSoup, Tag, NavigableString

logger = logging.getLogger(__name__)


def convert_content(
    xhtml: str,
    image_mapping: dict[str, str],
    image_folder: str,
) -> str:
    """Convert Confluence storage format XHTML to Flare body HTML.

    Args:
        xhtml: Raw XHTML from Confluence storage format.
        image_mapping: {confluence_filename: local_filename} from image handler.
        image_folder: Date-based folder name for image paths, e.g. "05-Jan-2026".

    Returns:
        Clean HTML string suitable for insertion into Flare topic <body>.
    """
    soup = BeautifulSoup(xhtml, "html.parser")

    _convert_images(soup, image_mapping, image_folder)
    _convert_macros(soup)
    _unwrap_layouts(soup)
    _demote_h1_to_h2(soup)
    _strip_attributes(soup)
    _remove_empty_paragraphs(soup)

    return _serialize(soup)


def _convert_images(
    soup: BeautifulSoup,
    image_mapping: dict[str, str],
    image_folder: str,
) -> None:
    """Replace ac:image elements with standard <img> tags."""
    for ac_image in soup.find_all("ac:image"):
        # Find the attachment reference inside
        ri_attachment = ac_image.find("ri:attachment")
        if ri_attachment:
            confluence_name = ri_attachment.get("ri:filename", "")
            local_name = image_mapping.get(confluence_name, confluence_name)
            src = f"../../Resources/From Confluence/Release Notes/{image_folder}/{local_name}"

            img_tag = soup.new_tag("img", src=src)
            # Preserve any width/height from the ac:image
            if ac_image.get("ac:width"):
                img_tag["style"] = f"width: {ac_image['ac:width']}px;"

            ac_image.replace_with(img_tag)
        else:
            # External image URL
            ri_url = ac_image.find("ri:url")
            if ri_url:
                url = ri_url.get("ri:value", "")
                img_tag = soup.new_tag("img", src=url)
                ac_image.replace_with(img_tag)
            else:
                ac_image.decompose()


def _convert_macros(soup: BeautifulSoup) -> None:
    """Convert ac:structured-macro elements to Flare equivalents."""
    for macro in soup.find_all("ac:structured-macro"):
        macro_name = macro.get("ac:name", "")

        if macro_name in ("info", "note"):
            _macro_to_callout(soup, macro, "note", "Note: ")
        elif macro_name == "warning":
            _macro_to_callout(soup, macro, "warning", "Warning: ")
        elif macro_name == "tip":
            _macro_to_callout(soup, macro, "tip", "Tip: ")
        elif macro_name == "code" or macro_name == "noformat":
            _macro_to_code(soup, macro)
        elif macro_name == "toc":
            # Remove TOC macros — Flare handles TOC differently
            macro.decompose()
        elif macro_name == "expand":
            _macro_expand(soup, macro)
        elif macro_name in ("panel", "section", "column"):
            # Unwrap panel/section/column — keep content only
            _unwrap_macro_body(macro)
        else:
            # Unknown macro — unwrap to keep content
            logger.debug("Unknown macro '%s', unwrapping", macro_name)
            _unwrap_macro_body(macro)


def _macro_to_callout(
    soup: BeautifulSoup, macro: Tag, css_class: str, prefix: str
) -> None:
    """Convert info/warning/tip macros to Flare callout divs."""
    body = macro.find("ac:rich-text-body")
    div = soup.new_tag("div")
    div["class"] = css_class
    div["MadCap:autonum"] = f"<b>{prefix}</b>"

    if body:
        for child in list(body.children):
            div.append(child.extract())

    macro.replace_with(div)


def _macro_to_code(soup: BeautifulSoup, macro: Tag) -> None:
    """Convert code/noformat macros to <pre><code>."""
    body = macro.find("ac:plain-text-body")
    text = body.string if body else ""

    pre = soup.new_tag("pre")
    code = soup.new_tag("code")
    code.string = text or ""
    pre.append(code)
    macro.replace_with(pre)


def _macro_expand(soup: BeautifulSoup, macro: Tag) -> None:
    """Convert expand macros — keep the body content, add title as heading."""
    title_param = macro.find("ac:parameter", attrs={"ac:name": "title"})
    body = macro.find("ac:rich-text-body")

    container = soup.new_tag("div")
    container["class"] = "expandable"

    if title_param and title_param.string:
        h3 = soup.new_tag("h3")
        h3.string = title_param.string
        container.append(h3)

    if body:
        for child in list(body.children):
            container.append(child.extract())

    macro.replace_with(container)


def _unwrap_macro_body(macro: Tag) -> None:
    """Unwrap a macro, keeping only its rich-text-body children."""
    body = macro.find("ac:rich-text-body")
    if body:
        macro.replace_with(body)
        body.unwrap()
    else:
        macro.decompose()


def _unwrap_layouts(soup: BeautifulSoup) -> None:
    """Remove Confluence layout wrappers, keeping child content."""
    for tag_name in ["ac:layout", "ac:layout-section", "ac:layout-cell"]:
        for tag in soup.find_all(tag_name):
            tag.unwrap()


def _demote_h1_to_h2(soup: BeautifulSoup) -> None:
    """Demote H1 headings to H2 (the Flare template provides its own H1)."""
    for h1 in soup.find_all("h1"):
        h1.name = "h2"


def _strip_attributes(soup: BeautifulSoup) -> None:
    """Strip Confluence-specific and style attributes from all elements."""
    KEEP_ATTRS = {"src", "href", "class", "id", "colspan", "rowspan",
                  "MadCap:autonum", "alt", "title", "width", "height"}

    for tag in soup.find_all(True):
        attrs_to_remove = []
        for attr_name in list(tag.attrs.keys()):
            # Remove Confluence-specific attributes
            if attr_name.startswith(("ac:", "ri:", "data-", "local-id")):
                attrs_to_remove.append(attr_name)
            # Remove inline styles (except on img tags where we set width)
            elif attr_name == "style" and tag.name != "img":
                attrs_to_remove.append(attr_name)
            # Remove any attribute not in the keep list
            elif attr_name not in KEEP_ATTRS:
                attrs_to_remove.append(attr_name)

        for attr_name in attrs_to_remove:
            del tag[attr_name]


def _remove_empty_paragraphs(soup: BeautifulSoup) -> None:
    """Remove empty <p> tags that are just whitespace."""
    for p in soup.find_all("p"):
        if not p.get_text(strip=True) and not p.find_all(["img", "br"]):
            p.decompose()


def _serialize(soup: BeautifulSoup) -> str:
    """Serialize the soup back to an HTML string, cleaning up whitespace."""
    html = str(soup)
    # Clean up Confluence namespace remnants
    html = re.sub(r"</?ac:[^>]+>", "", html)
    html = re.sub(r"</?ri:[^>]+>", "", html)
    # Clean up multiple blank lines
    html = re.sub(r"\n{3,}", "\n\n", html)
    return html.strip()

File: swift-app/ConfluenceToFlare/Services/ContentConverter.swift
import Foundation
import SwiftSoup

/// Converts Confluence storage format XHTML to Flare-compatible HTML.
///
/// Handles Confluence-specific elements (ac:image, ac:structured-macro,
/// ac:layout, ri:attachment, etc.) and converts them to clean HTML that
/// works in MadCap Flare topics.
struct ContentConverter {

    /// Convert Confluence storage format XHTML to Flare body HTML.
    ///
    /// - Parameters:
    ///   - xhtml: Raw XHTML from Confluence storage format.
    ///   - imageMapping: {confluence_filename: local_filename} from image handler.
    ///   - imageFolder: Date-based folder name for image paths, e.g. "05-Jan-2026".
    /// - Returns: Clean HTML string suitable for insertion into Flare topic body.
    static func convert(
        xhtml: String,
        imageMapping: [String: String],
        imageFolder: String
    ) throws -> String {
        // Pre-extract ac:width values before SwiftSoup parsing (HTML5 parser drops ac: attributes)
        let imageWidths = extractImageWidths(from: xhtml)

        let doc = try SwiftSoup.parseBodyFragment(xhtml)

        try convertImages(doc: doc, imageMapping: imageMapping, imageFolder: imageFolder, imageWidths: imageWidths)
        try convertMacros(doc: doc)
        try unwrapLayouts(doc: doc)
        try demoteH1ToH2(doc: doc)
        try stripAttributes(doc: doc)
        try removeEmptyParagraphs(doc: doc)

        return try serialize(doc: doc)
    }

    /// Convert for preview display, using direct Confluence URLs for images.
    ///
    /// - Parameters:
    ///   - xhtml: Raw XHTML from Confluence storage format.
    ///   - imageURLs: {confluence_filename: full_download_url} for remote image display.
    /// - Returns: Clean HTML string with images pointing to Confluence.
    static func convertForPreview(
        xhtml: String,
        imageURLs: [String: String]
    ) throws -> String {
        let imageWidths = extractImageWidths(from: xhtml)

        let doc = try SwiftSoup.parseBodyFragment(xhtml)

        try convertImagesForPreview(doc: doc, imageURLs: imageURLs, imageWidths: imageWidths)
        try convertMacros(doc: doc)
        try unwrapLayouts(doc: doc)
        try demoteH1ToH2(doc: doc)
        try stripAttributes(doc: doc)
        try removeEmptyParagraphs(doc: doc)

        return try serialize(doc: doc)
    }

    // MARK: - Pre-extraction (workaround for SwiftSoup dropping ac: attributes)

    /// Extract ac:width values from raw XHTML keyed by ri:filename.
    ///
    /// SwiftSoup's HTML5 parser strips `ac:` namespace attributes, so we
    /// pre-extract them with regex before parsing. Maps filenames to widths.
    private static func extractImageWidths(from xhtml: String) -> [String: String] {
        var widths: [String: String] = [:]

        // Match ac:image blocks and extract width + filename
        // Pattern handles self-closing ri:attachment (with />) and full closing tags
        guard let pattern = try? NSRegularExpression(
            pattern: "<ac:image[^>]*?ac:width=\"(\\d+)\"[^>]*>.*?ri:filename=\"([^\"]+)\".*?</ac:image>",
            options: [.dotMatchesLineSeparators]
        ) else { return widths }

        let matches = pattern.matches(in: xhtml, range: NSRange(xhtml.startIndex..., in: xhtml))
        for match in matches {
            guard let widthRange = Range(match.range(at: 1), in: xhtml),
                  let filenameRange = Range(match.range(at: 2), in: xhtml) else { continue }
            widths[String(xhtml[filenameRange])] = String(xhtml[widthRange])
        }

        return widths
    }

    // MARK: - Image Conversion

    /// Replace ac:image elements with standard <img> tags.
    private static func convertImages(
        doc: Document,
        imageMapping: [String: String],
        imageFolder: String,
        imageWidths: [String: String]
    ) throws {
        for acImage in try doc.getElementsByTag("ac:image").array() {
            if let riAttachment = try acImage.getElementsByTag("ri:attachment").first() {
                let confluenceName = try riAttachment.attr("ri:filename")
                let localName = imageMapping[confluenceName] ?? confluenceName
                let src = "../../Resources/From Confluence/Release Notes/\(imageFolder)/\(localName)"

                let imgTag = try doc.createElement("img")
                try imgTag.attr("src", src)

                // Use pre-extracted width (SwiftSoup drops ac: attributes during HTML5 parsing)
                if let width = imageWidths[confluenceName], !width.isEmpty {
                    try imgTag.attr("style", "width: \(width)px;")
                }

                try acImage.replaceWith(imgTag)
            } else if let riURL = try acImage.getElementsByTag("ri:url").first() {
                let url = try riURL.attr("ri:value")
                let imgTag = try doc.createElement("img")
                try imgTag.attr("src", url)
                try acImage.replaceWith(imgTag)
            } else {
                try acImage.remove()
            }
        }
    }

    /// Replace ac:image elements with <img> tags using direct Confluence URLs (for preview).
    private static func convertImagesForPreview(
        doc: Document,
        imageURLs: [String: String],
        imageWidths: [String: String]
    ) throws {
        for acImage in try doc.getElementsByTag("ac:image").array() {
            if let riAttachment = try acImage.getElementsByTag("ri:attachment").first() {
                let confluenceName = try riAttachment.attr("ri:filename")

                let imgTag = try doc.createElement("img")
                if let url = imageURLs[confluenceName] {
                    try imgTag.attr("src", url)
                } else {
                    try imgTag.attr("alt", "[Image: \(confluenceName)]")
                }

                if let width = imageWidths[confluenceName], !width.isEmpty {
                    try imgTag.attr("style", "width: \(width)px;")
                }

                try acImage.replaceWith(imgTag)
            } else if let riURL = try acImage.getElementsByTag("ri:url").first() {
                let url = try riURL.attr("ri:value")
                let imgTag = try doc.createElement("img")
                try imgTag.attr("src", url)
                try acImage.replaceWith(imgTag)
            } else {
                try acImage.remove()
            }
        }
    }

    // MARK: - Macro Conversion

    /// Convert ac:structured-macro elements to Flare equivalents.
    private static func convertMacros(doc: Document) throws {
        for macro in try doc.getElementsByTag("ac:structured-macro").array() {
            let macroName = try macro.attr("ac:name")

            switch macroName {
            case "info", "note":
                try macroToCallout(doc: doc, macro: macro, cssClass: "note", prefix: "Note: ")
            case "warning":
                try macroToCallout(doc: doc, macro: macro, cssClass: "warning", prefix: "Warning: ")
            case "tip":
                try macroToCallout(doc: doc, macro: macro, cssClass: "tip", prefix: "Tip: ")
            case "code", "noformat":
                try macroToCode(doc: doc, macro: macro)
            case "toc":
                try macro.remove()
            case "expand":
                try macroExpand(doc: doc, macro: macro)
            case "panel", "section", "column":
                try unwrapMacroBody(macro: macro)
            default:
                try unwrapMacroBody(macro: macro)
            }
        }
    }

    /// Convert info/warning/tip macros to Flare callout divs.
    private static func macroToCallout(
        doc: Document,
        macro: Element,
        cssClass: String,
        prefix: String
    ) throws {
        let body = try macro.getElementsByTag("ac:rich-text-body").first()
        let div = try doc.createElement("div")
        try div.attr("class", cssClass)
        try div.attr("MadCap:autonum", "<b>\(prefix)</b>")

        if let body {
            for child in body.getChildNodes() {
                try div.appendChild(child)
            }
        }

        try macro.replaceWith(div)
    }

    /// Convert code/noformat macros to <pre><code>.
    private static func macroToCode(doc: Document, macro: Element) throws {
        let body = try macro.getElementsByTag("ac:plain-text-body").first()
        let text = body != nil ? (try body!.text()) : ""

        let pre = try doc.createElement("pre")
        let code = try doc.createElement("code")
        try code.text(text)
        try pre.appendChild(code)
        try macro.replaceWith(pre)
    }

    /// Convert expand macros — keep body content, add title as heading.
    private static func macroExpand(doc: Document, macro: Element) throws {
        let container = try doc.createElement("div")
        try container.attr("class", "expandable")

        // Look for title parameter
        for param in try macro.getElementsByTag("ac:parameter").array() {
            let paramName = try param.attr("ac:name")
            if paramName == "title" {
                let titleText = try param.text()
                if !titleText.isEmpty {
                    let h3 = try doc.createElement("h3")
                    try h3.text(titleText)
                    try container.appendChild(h3)
                }
                break
            }
        }

        // Move body content
        if let body = try macro.getElementsByTag("ac:rich-text-body").first() {
            for child in body.getChildNodes() {
                try container.appendChild(child)
            }
        }

        try macro.replaceWith(container)
    }

    /// Unwrap a macro, keeping only its rich-text-body children.
    private static func unwrapMacroBody(macro: Element) throws {
        if let body = try macro.getElementsByTag("ac:rich-text-body").first() {
            // Move body's children to before the macro, then remove macro
            let parent = macro.parent()
            if let parent {
                let children = body.getChildNodes()
                let macroIndex = macro.siblingIndex
                for (i, child) in children.enumerated() {
                    try parent.insertChildren(macroIndex + i, [child])
                }
            }
            try macro.remove()
        } else {
            try macro.remove()
        }
    }

    // MARK: - Layout Unwrapping

    /// Remove Confluence layout wrappers, keeping child content.
    private static func unwrapLayouts(doc: Document) throws {
        for tagName in ["ac:layout", "ac:layout-section", "ac:layout-cell"] {
            // Process iteratively since unwrapping changes the tree
            while true {
                guard let tag = try doc.getElementsByTag(tagName).first() else { break }
                // Move children out, then remove the wrapper
                let parent = tag.parent()
                if let parent {
                    let tagIndex = tag.siblingIndex
                    let children = tag.getChildNodes()
                    for (i, child) in children.enumerated() {
                        try parent.insertChildren(tagIndex + i, [child])
                    }
                }
                try tag.remove()
            }
        }
    }

    // MARK: - H1 Demotion

    /// Demote H1 headings to H2 (the Flare template provides its own H1).
    private static func demoteH1ToH2(doc: Document) throws {
        for h1 in try doc.getElementsByTag("h1").array() {
            try h1.tagName("h2")
        }
    }

    // MARK: - Attribute Stripping

    private static let keepAttributes: Set<String> = [
        "src", "href", "class", "id", "colspan", "rowspan",
        "MadCap:autonum", "alt", "title", "width", "height",
    ]

    /// Strip Confluence-specific and style attributes from all elements.
    private static func stripAttributes(doc: Document) throws {
        for tag in try doc.getAllElements().array() {
            let attributes = tag.getAttributes()
            guard let attributes else { continue }

            var toRemove: [String] = []
            for attr in attributes {
                let key = attr.getKey()
                if key.hasPrefix("ac:") || key.hasPrefix("ri:") ||
                   key.hasPrefix("data-") || key == "local-id" {
                    toRemove.append(key)
                } else if key == "style" {
                    // Keep style only on img tags (where we set width from ac:width)
                    if tag.tagName() != "img" {
                        toRemove.append(key)
                    }
                } else if !keepAttributes.contains(key) {
                    toRemove.append(key)
                }
            }

            for key in toRemove {
                try tag.removeAttr(key)
            }
        }
    }

    // MARK: - Empty Paragraph Removal

    /// Remove empty <p> tags that are just whitespace.
    private static func removeEmptyParagraphs(doc: Document) throws {
        for p in try doc.getElementsByTag("p").array() {
            let text = try p.text().trimmingCharacters(in: .whitespacesAndNewlines)
            let hasImgOrBr = try !p.getElementsByTag("img").isEmpty() || !p.getElementsByTag("br").isEmpty()
            if text.isEmpty && !hasImgOrBr {
                try p.remove()
            }
        }
    }

    // MARK: - Serialization

    /// Serialize the document back to an HTML string, cleaning up namespace remnants.
    private static func serialize(doc: Document) throws -> String {
        var html = try doc.body()?.html() ?? ""

        // Restore MadCap attribute casing (SwiftSoup lowercases it)
        html = html.replacingOccurrences(of: "madcap:autonum", with: "MadCap:autonum")

        // Clean up Confluence namespace remnants
        html = html.replacingOccurrences(
            of: "</?ac:[^>]+>",
            with: "",
            options: .regularExpression
        )
        html = html.replacingOccurrences(
            of: "</?ri:[^>]+>",
            with: "",
            options: .regularExpression
        )

        // Clean up multiple blank lines
        html = html.replacingOccurrences(
            of: "\n{3,}",
            with: "\n\n",
            options: .regularExpression
        )

        return html.trimmingCharacters(in: .whitespacesAndNewlines)
    }
}
